fiftyone/utils/multiprocessing.py

```diff
@@ -0,0 +1,572 @@
+            for sample in batch_view.iter_samples(autosave=save):
+                outputs[sample.id] = map_fcn(sample)
+
+        output = aggregate_fcn(sample_collection, outputs)
+
+    Example::
+
+        from collections import Counter
+
+        import fiftyone as fo
+        import fiftyone.utils.multiprocessing as foum
+        import fiftyone.zoo as foz
+
+        dataset = foz.load_zoo_dataset("cifar10", split="train")
+        view = dataset.select_fields("ground_truth")
+
+        #
+        # Example 1: update
+        #
+
+        def map_fcn(sample):
+            sample.ground_truth.label = sample.ground_truth.label.upper()
+
+        foum.map_samples(view, map_fcn, return_outputs=False, save=True)
+
+        print(dataset.count_values("ground_truth.label"))
+
+        #
+        # Example 2: map
+        #
+
+        def map_fcn(sample):
+            return sample.ground_truth.label.lower()
+
+        counter = Counter()
+        for _, label in foum.map_samples(view, map_fcn):
+            counter[label] += 1
+
+        print(dict(counter))
+
+        #
+        # Example 3: map-reduce
+        #
+
+        def map_fcn(sample):
+            return sample.ground_truth.label.lower()
+
+        class ReduceFcn(fo.ReduceFcn):
+            def init(self):
+                self.accumulator = Counter()
+
+            def add(self, sample_id, output):
+                self.accumulator[output] += 1
+
+            def finalize(self):
+                return dict(self.accumulator)
+
+        counts = foum.map_samples(view, map_fcn, reduce_fcn=ReduceFcn)
+        print(counts)
+
+        #
+        # Example 4: map-aggregate
+        #
+
+        def map_fcn(sample):
+            return sample.ground_truth.label.lower()
+
+        def aggregate_fcn(sample_collection, outputs):
+            return dict(Counter(outputs.values()))
+
+        counts = foum.map_samples(view, map_fcn, aggregate_fcn=aggregate_fcn)
+        print(counts)
+
+    Args:
+        sample_collection: a
+            :class:`fiftyone.core.collections.SampleCollection`
+        map_fcn: a function to apply to each sample
+        reduce_fcn (None): an optional :class:`ReduceFcn` subclass to reduce
+            the map outputs. See above for usage information
+        aggregate_fcn (None): an optional function to aggregate the map
+            outputs. See above for usage information
+        return_outputs (True): whether to return the map outputs. This has no
+            effect when a ``reduce_fcn`` or ``aggregate_fcn`` is provided
+        save (False): whether to save any sample edits applied by ``map_fcn``
+        num_workers (None): the number of workers to use. The default is
+            :meth:`fiftyone.core.utils.recommend_process_pool_workers`. If this
+            value is <= 1, all work is done in the main process
+        shard_size (None): an optional number of samples to distribute to each
```

Comment:
I think we should consider removing shard_size since it is confusing to the users. If num_workers * shard_size < n, we will just end up using max_shard_size anyway :)  so might as well not need it